### DISCUSSION TITLE
Topology Optimization of Heat Transfer Problems

### DISCUSSION THREAD

**Discussion by schewef** [mdolab/dafoam#127] [General]
Dear @friedenhe 

as you know I am interested in topology optimization unsing DAFaom.

Especially I am interested in heat transfer problems. Therefore I would like to ask the following questions:

1. As part of the repository the heat transfer solver "SimpleTFoam" is available. According to your publication you already investigated more solvers like "buoyantSimpleFoam". Is it planned to include that solver to the repository? 

![image](https://user-images.githubusercontent.com/43476213/114593635-98edae00-9c8c-11eb-9edb-5e1a07016002.png)

2. For topology optimization we need a source term for the porosity in e.g. simpleTFoam which is currently not implemented (as far as I know). The implementation in DAFoam itself should be straight forward. What I am wondering about is, if the porosity term has implications on the AD version of OpenFOAM? Would one have to change/extend the AD-OpenFOAM at all or is the calculation of sensitivities according to the porosity independent from the solver implementation in DAFoam?

**Comment by friedenhe** [comment]
I just added the alphaPorosity term for DASimpleTFoam. Check this https://github.com/mdolab/dafoam/commit/9b0f251a12e9f097b51b0313d67cc4a52085857e. It should work as is; no need to change the AD version of OpenFOAM or the calculation of alpha sensitivity in DAFoam.

**Reply by schewef** [reply]
Great, thank you very much!

**Reply by friedenhe** [reply]
@schewef, we forgot to add alphaPorosity for the DAResidual class in the previous commit. Use this one instead https://github.com/mdolab/dafoam/commit/4fca748dd686ac8a7a7815d423fc594028c4ffb6

**Reply by schewef** [reply]
Dear @friedenhe 

I would like to replace the alphaPorosity term by another one for the relative density (design varibale between 0 and 1 in topology optimization) to directly calculate the sensitivities for the relative density.

I already managed to implement the modified term in the DASimpleTFoam module and to recompile dafoam with this modification. The solving works fine.

Problems occure when it comes to calculating the sensitivities. 
 
Generally spoken I just modified the term below in the files "UEqnSimpleT.H" and "DAResidualSimpleTFoam.C".

![image](https://user-images.githubusercontent.com/43476213/133788305-d2356049-4b62-454d-ae1b-469ae296acf6.png)

One could e.g. replace it with something like that (just as an simple example):

![image](https://user-images.githubusercontent.com/43476213/133789799-255e4eb6-7f69-4908-b4cf-7d229098dc6c.png)

If I complile and run dafoam with that example I would expect that the sensitivity field looks different from the sensitivity field of the unmodified solver. But this actually is not the case. They are still exactly the same.

Is it generally possible to define a new term in dafoam different from the ones that are implemented?
Do you have to recompile the AD Version of OpenFOAM after changing the term in DAFoam to get the correct sensitivities or do these two modules work independet from each other?
Or am I on a completely wrong track?

Thanks in advance!

**Reply by friedenhe** [reply]
Yes, they should be different. Are you sure you re-compiled both original and AD versions of DAFoam and run `pip install .`? BTW: no need to recompile the original or AD versions of OpenFOAM.

**Reply by schewef** [reply]
Thank you, I got it! Something went wrong with compiling the AD version of DAFoam. 
I also noted that the "Compiling from Source" documentation was updated. It's much clearer now!

**Reply by friedenhe** [reply]
We just updated the version to 2.2.9. This update fixed an issue in AD communication and improved the adjoint accuracy.

**Reply by chairmanmao256** [reply]
@schewef How did you check the sensitivity field? I wonder if there's a way to visualize the sensitivity field in Paraview or Tecplot.

**Reply by friedenhe** [reply]
Check this [page](https://dafoam.github.io/mydoc_verifications_overview.html) and the tutorial.

**Reply by chairmanmao256** [reply]
@friedenhe Thank you! However, I'm working on Topology optimization these days, the dimension of design variable (alphaporosity) is as large as the number of grid cells. So I wonder if there's a way to visualize this sensitivity 'field' in Paraview or something else.

**Reply by friedenhe** [reply]
https://dafoam.github.io/mydoc_get_started_faq.html#how-to-get-sensitivity-maps

**Reply by chairmanmao256** [reply]
@friedenhe Thank you! I should read DAFoam's website more carefully!

**Comment by sodabryant** [comment]


**Reply by friedenhe** [reply]
1. HFL0 is the heat flux for the baseline design. You can get this information from the first flow solution
2. Right, it is not consistent... We have not implemented the Nusselt number in DAFoam version 2 yet... The thing we have not added to version 2 is a variable dT that normalizes the heat flux. 
3. The reason to include only a portion of the channel is to be consistent with the experiment. Check the reference in the IJHMT paper.

**Reply by sodabryant** [reply]
I got it!!
Thanks and regards.

**Comment by sodabryant** [comment]
Dear @friedenhe 
Is it true that the topology optimization case of pitzDaily only works with the SNOPT optimizer? To solve the TO problems, any alternative optimizer do we have ?

**Reply by friedenhe** [reply]
IPOPT may also work, but we have not managed to make it yet...
